### ISSUE TITLE
Gradient Ascent  Teddy Koker

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#5192]
Gradient Ascent · Teddy Koker<br>
<br>
https://ift.tt/2TJzFrk<br>
<br>
About 0 Minutes<br>
<br>
<div><div>
<div>
<time datetime="2019-05-28T08:06:21-0400">Tue, May 28, 2019</time>
<p>In the next few posts, I will be going over a strategy that uses Machine Learning to determine what trades to execute. Before we start going over the strategy, we will go over one of the algorithms it uses: Gradient Ascent.</p>
<h2>What is Gradient Ascent?</h2>
<p>Gradient ascent is an algorithm used to maximize a given reward function. A common method to describe gradient ascent uses the following scenario: Imagine you are blindfolded and placed somewhere on a mountain. Your task is then to find the highest point of the mountain. In this scenario, the “reward function” you are trying to maximize is your elevation. A simple method of locating this maximum would be observing the slope of the area you are standing on, and moving uphill. Following these directions one step at a time would eventually lead you to the top!</p>
<p>While making our way up the mountain, it was important that we knew the slope, or gradient, of the area, so we could know which direction to travel. This <code>gradient</code> is simply the derivative of the reward function with respect to its parameters.</p>
<p>Another important piece of gradient ascent is <code>learning rate</code>. This is equivalent to the number of steps we take before checking the slope again. Too many steps, and we could overshoot the summit; too few steps, and finding the peak would take way too long. Similarly, a high <code>learning rate</code> may lead the algorithm to diverge from the maximum, while a low <code>learing rate</code> might result in the algorithm taking too long to finish.</p>
<p>Now that we understand the basics of gradient ascent, let’s use it to perform a relatively simple task: linear regression.</p>
<h2>Example: Linear Regression using Gradient Ascent</h2>
<p>Let’s first generate some data to perform linear regression with.</p>
<div><pre><code data-lang="python"><span>%</span><span>matplotlib</span> <span>inline</span>
<span>import</span> <span>numpy</span> <span>as</span> <span>np</span>
<span>import</span> <span>matplotlib.pyplot</span> <span>as</span> <span>plt</span>
<span>plt</span><span>.</span><span>rcParams</span><span>[</span><span>"figure.figsize"</span><span>]</span> <span>=</span> <span>(</span><span>5</span><span>,</span> <span>3</span><span>)</span> <span># (w, h)</span>
<span>plt</span><span>.</span><span>rcParams</span><span>[</span><span>"figure.dpi"</span><span>]</span> <span>=</span> <span>200</span></code></pre></div>
<div><pre><code data-lang="python"><span>m</span> <span>=</span> <span>100</span>
<span>x</span> <span>=</span> <span>2</span> <span>*</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>rand</span><span>(</span><span>m</span><span>)</span>
<span>y</span> <span>=</span> <span>5</span> <span>+</span> <span>2</span> <span>*</span> <span>x</span> <span>+</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>randn</span><span>(</span><span>m</span><span>)</span>
<span>plt</span><span>.</span><span>scatter</span><span>(</span><span>x</span><span>,</span> <span>y</span><span>);</span></code></pre></div>
<p><img alt="png" src="https://teddykoker.com/2019/05/trading-with-reinforcement-learning-in-python-part-i-gradient-ascent/output_9_0.png"></p>
<p>We generated 100 points placed randomly around a line with an intercept of 5, and a slope of 2. Just so we have a benchmark, we can find the line of best fit using scipy’s <code>lingregress</code> function:</p>
<div><pre><code data-lang="python"><span>from</span> <span>scipy.stats</span> <span>import</span> <span>linregress</span>
<span>slope</span><span>,</span> <span>intercept</span> <span>=</span> <span>linregress</span><span>(</span><span>x</span><span>,</span> <span>y</span><span>)[:</span><span>2</span><span>]</span>
<span>print</span><span>(</span><span>f</span><span>"slope: {slope:.3f}, intercept: {intercept:.3f}"</span><span>)</span></code></pre></div>
<pre><code>slope: 2.064, intercept: 4.889
</code></pre>
<p>These will be the values to shoot for with our gradient ascent algorithm.</p>
<h3>Reward Function</h3>
<p>The next step is to define our reward function. A commonly used function when measuring the accuracy of linear regression is <a href="https://en.wikipedia.org/wiki/Mean_squared_error">Mean Squared Error</a> (MSE). MSE is the “average squared difference between the estimated values and what is estimated.” Since MSE is an <em>error</em> function, and we are looking for a function to maximized we will use negative MSE as our reward function, <span><span><span></span><span><span><span></span><span>J</span></span></span></span></span>:</p>
<p><span><span><span><span></span><span><span><span></span><span>J</span><span>(</span><span>θ</span><span>)</span><span></span><span>=</span><span></span></span><span><span></span><span>−</span><span><span><span></span><span><span><span><span><span><span></span><span><span>m</span></span></span><span><span></span><span></span></span><span><span></span><span><span>1</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span><span></span></span></span><span></span><span><span><span><span><span><span></span><span><span><span>i</span><span>=</span><span>1</span></span></span></span><span><span></span><span><span>∑</span></span></span><span><span></span><span><span><span>m</span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span><span>(</span><span><span>θ</span><span><span><span><span><span><span></span><span><span><span>0</span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span><span></span><span>+</span><span></span></span><span><span></span><span><span>θ</span><span><span><span><span><span><span></span><span><span><span>1</span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span><span><span>x</span><span><span><span><span><span><span></span><span><span><span>(</span><span>i</span><span>)</span></span></span></span></span></span></span></span></span><span></span><span>−</span><span></span></span><span><span></span><span><span>y</span><span><span><span><span><span><span></span><span><span><span>(</span><span>i</span><span>)</span></span></span></span></span></span></span></span></span><span><span>)</span><span><span><span><span><span><span></span><span><span>2</span></span></span></span></span></span></span></span></span></span></span></span></span></p>
<p>Where <span><span><span></span><span><span><span></span><span>θ</span></span></span></span></span> is our input parameters, in this case the intercept and slope of the line we are testing. This equation can be represented in Python like so:</p>
<div><pre><code data-lang="python"><span>x</span> <span>=</span> <span>np</span><span>.</span><span>array</span><span>([</span><span>np</span><span>.</span><span>ones</span><span>(</span><span>m</span><span>),</span> <span>x</span><span>])</span><span>.</span><span>transpose</span><span>()</span>
<span>def</span> <span>accuracy</span><span>(</span><span>x</span><span>,</span> <span>y</span><span>,</span> <span>theta</span><span>):</span>
    <span>return</span> <span>-</span> <span>1</span> <span>/</span> <span>m</span> <span>*</span> <span>np</span><span>.</span><span>sum</span><span>((</span><span>np</span><span>.</span><span>dot</span><span>(</span><span>x</span><span>,</span> <span>theta</span><span>)</span> <span>-</span> <span>y</span><span>)</span> <span>**</span> <span>2</span><span>)</span></code></pre></div>
<p>From here on I have made <span><span><span></span><span><span><span></span><span>x</span></span></span></span></span> a matrix where <span><span><span></span><span><span><span></span><span><span>x</span><span><span><span><span><span><span></span><span><span>0</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span><span></span><span>=</span><span></span></span><span><span></span><span>1</span></span></span></span></span> and <span><span><span></span><span><span><span></span><span><span>x</span><span><span><span><span><span><span></span><span><span>1</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span></span></span></span></span> is the original x values. This makes it so <span><span><span></span><span><span><span></span><span><span>θ</span><span><span><span><span><span><span></span><span><span><span>0</span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span><span></span><span>+</span><span></span></span><span><span></span><span><span>θ</span><span><span><span><span><span><span></span><span><span><span>1</span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span><span>x</span></span></span></span></span> can be computed with just <span><span><span></span><span><span><span></span><span>θ</span><span></span><span>⋅</span><span></span></span><span><span></span><span>x</span></span></span></span></span>.</p>
<h3>Gradient Function</h3>
<p>Now that we have our reward function, we can find our gradient function which will be the partial derivative of <span><span><span></span><span><span><span></span><span>J</span></span></span></span></span> with respect to <span><span><span></span><span><span><span></span><span>θ</span></span></span></span></span>:</p>
<p><span><span><span><span></span><span><span><span></span><span><span><span></span><span><span><span><span><span><span></span><span><span>∂</span><span>θ</span></span></span><span><span></span><span></span></span><span><span></span><span><span>∂</span><span>J</span><span>(</span><span>θ</span><span>)</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span><span></span></span></span><span></span><span>=</span><span></span></span><span><span></span><span>−</span><span><span><span></span><span><span><span><span><span><span></span><span><span>m</span></span></span><span><span></span><span></span></span><span><span></span><span><span>2</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span><span></span></span></span><span></span><span><span><span><span><span><span></span><span><span><span>i</span><span>=</span><span>1</span></span></span></span><span><span></span><span><span>∑</span></span></span><span><span></span><span><span><span>m</span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span><span>(</span><span><span>θ</span><span><span><span><span><span><span></span><span><span><span>0</span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span><span></span><span>+</span><span></span></span><span><span></span><span><span>θ</span><span><span><span><span><span><span></span><span><span><span>1</span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span><span><span>x</span><span><span><span><span><span><span></span><span><span>1</span></span></span><span><span></span><span><span><span>(</span><span>i</span><span>)</span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span><span></span><span>−</span><span></span></span><span><span></span><span><span>y</span><span><span><span><span><span><span></span><span><span><span>(</span><span>i</span><span>)</span></span></span></span></span></span></span></span></span><span>)</span><span></span><span>⋅</span><span></span></span><span><span></span><span><span>x</span><span><span><span><span><span><span></span><span><span><span>(</span><span>i</span><span>)</span></span></span></span></span></span></span></span></span></span></span></span></span></span></p>
<p>Once again we can write this function in Python:</p>
<div><pre><code data-lang="python"><span>def</span> <span>gradient</span><span>(</span><span>x</span><span>,</span> <span>y</span><span>,</span> <span>theta</span><span>):</span>
    <span>return</span> <span>-</span><span>1</span> <span>/</span> <span>m</span> <span>*</span> <span>x</span><span>.</span><span>T</span><span>.</span><span>dot</span><span>(</span><span>np</span><span>.</span><span>dot</span><span>(</span><span>x</span><span>,</span> <span>theta</span><span>)</span> <span>-</span> <span>y</span><span>)</span></code></pre></div>
<h3>Training</h3>
<p>Now we are ready to perform gradient ascent! We will initialize <span><span><span></span><span><span><span></span><span>θ</span></span></span></span></span> as <span><span><span></span><span><span><span></span><span>[</span><span>0</span><span>,</span><span></span><span>0</span><span>]</span></span></span></span></span>, and update it every epoch, or step using: <span><span><span></span><span><span><span></span><span>θ</span><span></span><span>=</span><span></span></span><span><span></span><span>θ</span><span></span><span>+</span><span></span></span><span><span></span><span>α</span><span><span><span></span><span><span><span><span><span><span></span><span><span><span>∂</span><span>θ</span></span></span></span><span><span></span><span></span></span><span><span></span><span><span><span>∂</span><span>J</span><span>(</span><span>θ</span><span>)</span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span><span></span></span></span></span></span></span></span>, where <span><span><span></span><span><span><span></span><span>α</span></span></span></span></span> is our learning rate.</p>
<div><pre><code data-lang="python"><span>num_epochs</span> <span>=</span> <span>500</span>
<span>learning_rate</span> <span>=</span> <span>0.01</span>

<span>def</span> <span>train</span><span>(</span><span>x</span><span>,</span> <span>y</span><span>):</span>
    <span>accs</span> <span>=</span> <span>[]</span>
    <span>thetas</span> <span>=</span> <span>[]</span>
    <span>theta</span> <span>=</span> <span>np</span><span>.</span><span>zeros</span><span>(</span><span>2</span><span>)</span>
    <span>for</span> <span>_</span> <span>in</span> <span>range</span><span>(</span><span>num_epochs</span><span>):</span>
        <span># keep track of accuracy and theta over time</span>
        <span>acc</span> <span>=</span> <span>accuracy</span><span>(</span><span>x</span><span>,</span> <span>y</span><span>,</span> <span>theta</span><span>)</span>
        <span>thetas</span><span>.</span><span>append</span><span>(</span><span>theta</span><span>)</span>
        <span>accs</span><span>.</span><span>append</span><span>(</span><span>acc</span><span>)</span>

        <span># update theta</span>
        <span>theta</span> <span>=</span> <span>theta</span> <span>+</span> <span>learning_rate</span> <span>*</span> <span>gradient</span><span>(</span><span>x</span><span>,</span> <span>y</span><span>,</span> <span>theta</span><span>)</span>

    <span>return</span> <span>theta</span><span>,</span> <span>thetas</span><span>,</span> <span>accs</span>

<span>theta</span><span>,</span> <span>thetas</span><span>,</span> <span>accs</span> <span>=</span> <span>train</span><span>(</span><span>x</span><span>,</span> <span>y</span><span>)</span>
<span>print</span><span>(</span><span>f</span><span>"slope: {theta[1]:.3f}, intercept: {theta[0]:.3f}"</span><span>)</span></code></pre></div>
<pre><code>slope: 2.770, intercept: 4.043
</code></pre>
<p>We matched the linear regression benchmark! If we graph the accuracy over time, we can see that the algorithm quickly converges to a maximum accuracy:</p>
<div><pre><code data-lang="python"><span>plt</span><span>.</span><span>plot</span><span>(</span><span>accs</span><span>)</span>
<span>plt</span><span>.</span><span>xlabel</span><span>(</span><span>'Epoch Number'</span><span>)</span>
<span>plt</span><span>.</span><span>ylabel</span><span>(</span><span>'Accuracy'</span><span>);</span></code></pre></div>
<p><img alt="png" src="https://teddykoker.com/2019/05/trading-with-reinforcement-learning-in-python-part-i-gradient-ascent/output_28_0.png"></p>
<p>Finally, if we project our reward function onto a 3D surface and mark our <span><span><span></span><span><span><span></span><span><span>θ</span><span><span><span><span><span><span></span><span><span>0</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span></span></span></span></span> and <span><span><span></span><span><span><span></span><span><span>θ</span><span><span><span><span><span><span></span><span><span>1</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span></span></span></span></span> over time, we can see our gradient ascent algorithm gradually finding its way to the maximum:</p>
<div><pre><code data-lang="python"><span>from</span> <span>mpl_toolkits.mplot3d</span> <span>import</span> <span>Axes3D</span>
<span>i</span> <span>=</span> <span>np</span><span>.</span><span>linspace</span><span>(</span><span>-</span><span>10</span><span>,</span> <span>20</span><span>,</span> <span>50</span><span>)</span>
<span>j</span> <span>=</span> <span>np</span><span>.</span><span>linspace</span><span>(</span><span>-</span><span>10</span><span>,</span> <span>20</span><span>,</span> <span>50</span><span>)</span>
<span>i</span><span>,</span> <span>j</span> <span>=</span> <span>np</span><span>.</span><span>meshgrid</span><span>(</span><span>i</span><span>,</span> <span>j</span><span>)</span>
<span>k</span> <span>=</span> <span>np</span><span>.</span><span>array</span><span>([</span><span>accuracy</span><span>(</span><span>x</span><span>,</span> <span>y</span><span>,</span> <span>th</span><span>)</span> <span>for</span> <span>th</span> <span>in</span> <span>zip</span><span>(</span><span>np</span><span>.</span><span>ravel</span><span>(</span><span>i</span><span>),</span> <span>np</span><span>.</span><span>ravel</span><span>(</span><span>j</span><span>))])</span><span>.</span><span>reshape</span><span>(</span><span>i</span><span>.</span><span>shape</span><span>)</span>
<span>fig</span> <span>=</span> <span>plt</span><span>.</span><span>figure</span><span>(</span><span>figsize</span><span>=</span><span>(</span><span>9</span><span>,</span><span>6</span><span>))</span>
<span>ax</span> <span>=</span> <span>fig</span><span>.</span><span>gca</span><span>(</span><span>projection</span><span>=</span><span>'3d'</span><span>)</span>
<span>ax</span><span>.</span><span>plot_surface</span><span>(</span><span>i</span><span>,</span> <span>j</span><span>,</span> <span>k</span><span>,</span> <span>alpha</span><span>=</span><span>0.2</span><span>)</span>
<span>ax</span><span>.</span><span>plot</span><span>([</span><span>t</span><span>[</span><span>0</span><span>]</span> <span>for</span> <span>t</span> <span>in</span> <span>thetas</span><span>],</span> <span>[</span><span>t</span><span>[</span><span>1</span><span>]</span> <span>for</span> <span>t</span> <span>in</span> <span>thetas</span><span>],</span> <span>accs</span><span>,</span> <span>marker</span><span>=</span><span>"o"</span><span>,</span> <span>markersize</span><span>=</span><span>3</span><span>,</span> <span>alpha</span><span>=</span><span>0.1</span><span>);</span>
<span>ax</span><span>.</span><span>set_xlabel</span><span>(</span><span>r</span><span>'$\theta_0$'</span><span>);</span> <span>ax</span><span>.</span><span>set_ylabel</span><span>(</span><span>r</span><span>'$\theta_1$'</span><span>)</span>
<span>ax</span><span>.</span><span>set_zlabel</span><span>(</span><span>"Accuracy"</span><span>);</span></code></pre></div>
<p><img alt="png" src="https://teddykoker.com/2019/05/trading-with-reinforcement-learning-in-python-part-i-gradient-ascent/output_30_0.png"></p>
<p>In this post we showed how gradient ascent can be used to maximize a relatively simple reward function with only two parameters. In the next post we will see how a reward function can be applied to a trading strategy in order to train a algorithmic trading model. As always, the notebooks for each post are available on my <a href="https://github.com/teddykoker/blog/tree/master/notebooks">Github</a>.</p>
</div>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via teddykoker.com https://teddykoker.com<br>
<br>
August 19, 2019 at 06:11PM
